README.md
# hostdir (temporary name)

Hosts a folder using http

#### Command usage (from `--help`):

```
hostdir v0.1.0.0

  -p    --port=                  What port to listen on (default 8080)
  -h    --host=                  What host to listen on (default 127.0.0.1)
  -H    --any-host               Use host '*'
  -r    --root=                  Root folder            (default .)
        --help                   Show command usage
        --404=                   404 page               (default 404.html)
  -V    --version, --ver         Show version
  -B    --no-buffer              Don't buffer output
  -l    --buffer-line, --buffer  Use line buffering (default)
  -b[]  --buffer-block[=]        Use block buffering

Valid arguments to --host:
- *   Any host, both IPv4 and IPv6
- *4  Any host, prefer IPv4
- *6  Any host, prefer IPv6
- !4  Any host, only IPv4
- !6  Any host, only IPv6
Any other value is treated as a normal hostname.
```


ChangeLog.md
# Changelog for hostdir

## Unreleased changes


app/Main.hs
{-# LANGUAGE LambdaCase #-}

module Main where

import           Control.Monad
import qualified Data.ByteString.Lazy      as BSL
import           Data.String
import qualified Data.Text                 as Text
import qualified Data.Text.Encoding        as Text
import qualified Data.Time.Clock           as Time
import qualified Data.Time.Format          as Time
import           Network.Wai
import           Network.HTTP.Types.Status
import qualified Network.Wai.Handler.Warp  as Warp
import           System.Console.GetOpt
import           System.Directory
import           System.Environment
import           System.Exit
import           System.FilePath
import           System.IO
import           Text.Pandoc

import Lib
import Opts
import Settings

first :: Monad m => [a] -> (a -> m Bool) -> m (Maybe a)
first []      _ = pure Nothing
first (a : b) f = f a >>= \case
  True  -> pure $ pure a
  False -> first b f

serveFile :: (Response -> IO w) -> [String] -> Status -> FilePath -> IO w
serveFile respond tags code res = case takeExtension res of
    ".md" -> do
      putStrLn $ " -> " ++ formattedTags ++ "<Markdown as HTML> " ++ res

      mdcont <- readFile res
      cont <- runIOorExplode $ do
        v <- readMarkdown (def { readerStandalone = True }) $ Text.pack mdcont
        writeHtml5String def v

      respond $ responseLBS
        code
        [("Content-Type", "text/html")]
        (BSL.fromStrict (Text.encodeUtf8 cont))
    _ -> do
      putStrLn $ " -> " ++ formattedTags ++ "<File> " ++ res
      respond $ responseFile
        code
        []
        res
        Nothing
  where
    formattedTags = tags >>= (++ " ")

app :: Settings -> Application
app settings req respond = do
  let pathInfoReq = pathInfo req
  let requestString = joinPath (Text.unpack <$> pathInfoReq)

  do
    time <- Time.getCurrentTime
    putStr $ Time.formatTime Time.defaultTimeLocale "[%F %T%3Q %Z] " time
  putStr $ "<" ++ show (remoteHost req) ++ "> "
  putStr requestString

  -- Refuse to fulfill request if `..` found in the path
  if ".." `elem` pathInfoReq then do
    putStrLn " -> [403]"
    respond $ responseLBS
      status403
      [("Content-Type", "text/plain")]
      "Forbidden (suspicious request)"
  else do
    a <- first (fmap ($ requestString) (hhostPath settings)) doesFileExist
    case a of
      Nothing  -> serveFile respond ["[404]"] status404 (hhost404 settings)
      Just res -> serveFile respond []        status200 res

printVersion :: IO ()
#ifdef RELEASE
printVersion = putStrLn $ "hostdir v" ++ CURRENT_PACKAGE_VERSION
#else
printVersion = putStrLn $ "hostdir v" ++ CURRENT_PACKAGE_VERSION ++ ", compiled " ++ __DATE__ ++ " " ++ __TIME__
#endif

main :: IO ()
main = do
  args <- getArgs

  settings <- case applyOpts args defaultSettings of
    Left  errs      -> putStrLn `mapM_` errs >> exitFailure
    Right (set, []) -> makeSettingsAbsolute set
    Right (_,   _ ) -> putStrLn "hostdir takes no non-option arguments" >> exitFailure

  when (hhostVer settings) $ do
    printVersion
    exitSuccess

  when (hhostHelp settings) $ do
    printVersion
    putStr   $ usageInfo "" optDescrs
    putStrLn ""
    putStrLn "Valid arguments to --host:"
    putStrLn "- *   Any host, both IPv4 and IPv6"
    putStrLn "- *4  Any host, prefer IPv4"
    putStrLn "- *6  Any host, prefer IPv6"
    putStrLn "- !4  Any host, only IPv4"
    putStrLn "- !6  Any host, only IPv6"
    putStrLn "Any other value is treated as a normal hostname."
    exitSuccess

  hSetBuffering stdout (hhostBuf settings)

  putStrLn $ "Hosting folder " ++ hhostRoot settings ++ " on " ++ ppHostPreference (hhostHost settings) ++ ":" ++ show (hhostPort settings)

  Warp.runSettings (Warp.setHost (hhostHost settings) (Warp.setPort (hhostPort settings) Warp.defaultSettings)) (app settings)


app/Settings.hs
module Settings where

import Data.Streaming.Network
import System.Directory
import System.FilePath
import System.IO

data Settings = Settings
  { hhostPort :: Int
  , hhostHost :: HostPreference
  , hhostRoot :: FilePath
  , hhostPath :: [FilePath -> FilePath]
  , hhost404  :: FilePath
  , hhostHelp :: Bool
  , hhostVer  :: Bool
  , hhostBuf  :: BufferMode
  }

defaultSettings :: Settings
defaultSettings = Settings
  { hhostPort = 8080
  , hhostHost = "127.0.0.1"
  , hhostRoot = "."
  , hhostPath = [id, (</> "index.html"), (</> "index.md"), (++ ".html"), (++ ".md")]
  , hhost404  = "404.html"
  , hhostHelp = False
  , hhostVer  = False
  , hhostBuf  = LineBuffering
  }

makeSettingsAbsolute :: Settings -> IO Settings
makeSettingsAbsolute set = do
  absRoot <- makeAbsolute (hhostRoot set)
  pure $ set
    { hhostPath = fmap ((absRoot </>) .) (hhostPath set)
    , hhost404 = absRoot </> hhost404 set
    , hhostRoot = absRoot
    }


app/Opts.hs
module Opts
  ( optDescrs
  , parseOpts
  , applyOpts
  ) where

import Control.Monad
import Data.String
import Data.Streaming.Network.Internal
import System.Console.GetOpt
import System.IO
import Text.Read

import Settings

-- Endomorphism in the Kleisli category generated by `Either String`
newtype Opt = Opt { runOpt :: Settings -> Either String Settings }
instance Semigroup Opt where
  Opt a <> Opt b = Opt (b >=> a)
instance Monoid Opt where
  mempty = Opt pure

optDescrs :: [OptDescr Opt]
optDescrs =
    [ Option "p" ["port"]                  portArg "What port to listen on (default 8080)"
    , Option "h" ["host"]                  hostArg "What host to listen on (default 127.0.0.1)"
    , Option "H" ["any-host"]              anyhArg "Use host '*'"
    , Option "r" ["root"]                  rootArg "Root folder            (default .)"
    , Option ""  ["help"]                  helpArg "Show command usage"
    , Option ""  ["404"]                   _404Arg "404 page               (default 404.html)"
    , Option "V" ["version", "ver"]        verArg  "Show version"
    , Option "B" ["no-buffer"]             nobArg  "Don't buffer output"
    , Option "l" ["buffer-line", "buffer"] blArg   "Use line buffering (default)"
    , Option "b" ["buffer-block"]          bbArg   "Use block buffering"
    ]
  where
    portArg, hostArg, anyhArg, rootArg, _404Arg, verArg, nobArg, blArg, bbArg :: ArgDescr Opt
    portArg = ReqArg portHandler                                                         ""
    hostArg = ReqArg (\host -> Opt (\set -> pure (set { hhostHost = fromString host }))) ""
    anyhArg = NoArg           (Opt (\set -> pure (set { hhostHost = HostAny })))
    rootArg = ReqArg (\root -> Opt (\set -> pure (set { hhostRoot = root })))            ""
    _404Arg = ReqArg (\_404 -> Opt (\set -> pure (set { hhost404  = _404 })))            ""
    helpArg = NoArg (Opt (\set -> pure (set { hhostHelp = True })))
    verArg  = NoArg (Opt (\set -> pure (set { hhostVer  = True })))
    nobArg  = NoArg (Opt (\set -> pure (set { hhostBuf  = NoBuffering })))
    blArg   = NoArg (Opt (\set -> pure (set { hhostBuf  = LineBuffering })))
    bbArg   = OptArg bbHandler ""

    bbHandler :: Maybe String -> Opt
    bbHandler Nothing = Opt (\set -> pure (set { hhostBuf = BlockBuffering Nothing }))
    bbHandler (Just bs) = Opt $ \set -> case readMaybe bs of
      Nothing -> Left $ "Block size doesn't parse: " ++ show bs
      Just bs -> pure $ set { hhostBuf = BlockBuffering (Just bs) }

    portHandler :: String -> Opt
    portHandler str = Opt $ \set -> case readMaybe str of
      Nothing            -> Left $ "Port doesn't parse: " ++ show str
      Just n | n < 0     -> Left $ "Port less than 0: " ++ show n
             | n > 65535 -> Left $ "Port greater than 65535: " ++ show n
             | otherwise -> pure $ set { hhostPort = n }

parseOpts :: [String] -> ([Opt], [String], [String])
parseOpts = getOpt Permute optDescrs

applyOpts :: [String] -> Settings -> Either [String] (Settings, [String])
applyOpts args set =
  let (perm, nonopts, errors) = parseOpts args
  in case errors of
    [] -> case runOpt (mconcat perm) set of
      Left err  -> Left [err]
      Right res -> Right (res, nonopts)
    errs -> Left errs
